plugin/buycourses/src/buy_course_plugin.class.php

```diff
@@ -3311,4 +3833,444 @@ private function updateServiceSaleStatus(
+    {
+        $couponItemTable = Database::get_main_table(self::TABLE_COUPON_ITEM);
+        if ($productType == self::PRODUCT_TYPE_COURSE) {
+            $itemTable = Database::get_main_table(TABLE_MAIN_COURSE);
+            $select = ['ci.product_id as id', 'it.title'];
+        } elseif ($productType == self::PRODUCT_TYPE_SESSION) {
+            $itemTable = Database::get_main_table(TABLE_MAIN_SESSION);
+            $select = ['ci.product_id as id', 'it.name'];
+        }
+
+        $couponFrom = "
+            $couponItemTable ci
+            INNER JOIN $itemTable it
+                on it.id = ci.product_id and ci.product_type = $productType
+        ";
+
+        return Database::select(
+            $select,
+            $couponFrom,
+            [
+                'where' => [
+                    'ci.coupon_id = ? ' => (int) $couponId,
+                ],
+            ]
+        );
+    }
+
+    /**
+     * Get the services of a coupon.
+     *
+     * @param string $couponId The coupon ID
+     *
+     * @return array The service data
+     */
+    private function getServicesCoupons($couponId)
+    {
+        $couponServiceTable = Database::get_main_table(self::TABLE_COUPON_SERVICE);
+        $serviceTable = Database::get_main_table(self::TABLE_SERVICES);
+
+        $couponFrom = "
+            $couponServiceTable cs
+            INNER JOIN $serviceTable s
+                on s.id = cs.service_id
+        ";
+
+        return Database::select(
+            ['cs.service_id as id', 's.name'],
+            $couponFrom,
+            [
+                'where' => [
+                    'cs.coupon_id = ? ' => (int) $couponId,
+                ],
+            ]
+        );
+    }
+
+    /**
+     * Get an array of coupons filtered by their status.
+     *
+     * @param int $status The coupon activation status
+     *
+     * @return array Coupons data
+     */
+    private function getDataCoupons($status = null)
+    {
+        $couponTable = Database::get_main_table(self::TABLE_COUPON);
+
+        if ($status != null) {
+            return Database::select(
+                ['*'],
+                $couponTable,
+                [
+                    'where' => [
+                        ' active = ? ' => (int) $status,
+                    ],
+                    'order' => 'id DESC',
+                ]
+            );
+        } else {
+            return Database::select(
+                ['*'],
+                $couponTable,
+                [
+                    'order' => 'id DESC',
+                ]
+            );
+        }
+    }
+
+    /**
+     * Get data of a coupon for a product (course or service) by the coupon ID.
+     *
+     * @param int $couponId    The coupon code code
+     * @param int $productType The product type
+     * @param int $productId   The product ID
+     *
+     * @return array The coupon data
+     */
+    private function getDataCoupon($couponId, $productType = null, $productId = null)
```

Comment:
`getCouponData` would be a better English name

Reply:
I see many other methods are called with the reverse order, so I'm leaving this like this.